monkey/common/credentials/credentials.py

```diff
@@ -1,10 +1,144 @@
+                for component in data["secrets"]
+            ]
+        )
+
+        return data
+
+    @staticmethod
+    def _build_credential_component(data: Mapping[str, Any]) -> ICredentialComponent:
+        try:
+            credential_component_type = CredentialComponentType[data["credential_type"]]
+        except KeyError as err:
+            raise InvalidCredentialsError(f"Unknown credential component type {err}")
+
+        credential_component_class = CREDENTIAL_COMPONENT_TYPE_TO_CLASS[credential_component_type]
+        credential_component_schema = CREDENTIAL_COMPONENT_TYPE_TO_CLASS_SCHEMA[
+            credential_component_type
+        ]
+
+        try:
+            return credential_component_class(**credential_component_schema.load(data))
+        except MarshmallowError as err:
+            raise InvalidCredentialComponentError(credential_component_class, str(err))
+
+    @pre_dump
+    def _serialize_credentials(
+        self, credentials: Credentials, **kwargs
+    ) -> Mapping[str, Sequence[Mapping[str, Any]]]:
+        data = {}
+
+        data["identities"] = tuple(
+            [
+                CredentialsSchema._serialize_credential_component(component)
+                for component in credentials.identities
+            ]
+        )
+        data["secrets"] = tuple(
+            [
+                CredentialsSchema._serialize_credential_component(component)
+                for component in credentials.secrets
+            ]
+        )
+
+        return data
+
+    @staticmethod
+    def _serialize_credential_component(
+        credential_component: ICredentialComponent,
+    ) -> Mapping[str, Any]:
+        credential_component_schema = CREDENTIAL_COMPONENT_TYPE_TO_CLASS_SCHEMA[
+            credential_component.credential_type
+        ]
+
+        return credential_component_schema.dump(credential_component)
 
 
 @dataclass(frozen=True)
 class Credentials:
     identities: Tuple[ICredentialComponent]
     secrets: Tuple[ICredentialComponent]
+
+    @staticmethod
+    def from_mapping(credentials: Mapping) -> Credentials:
+        try:
+            deserialized_data = CredentialsSchema().load(credentials)
+            return Credentials(**deserialized_data)
+        except (InvalidCredentialsError, InvalidCredentialComponentError) as err:
+            raise err
+        except MarshmallowError as err:
+            raise InvalidCredentialsError(str(err))
+
+    @staticmethod
+    def from_json(credentials: str) -> Credentials:
+        try:
+            deserialized_data = CredentialsSchema().loads(credentials)
+            return Credentials(**deserialized_data)
+        except (InvalidCredentialsError, InvalidCredentialComponentError) as err:
+            raise err
+        except MarshmallowError as err:
+            raise InvalidCredentialsError(str(err))
+
+    @staticmethod
+    def to_json(credentials: Credentials) -> str:
+        return CredentialsSchema().dumps(credentials)
```

Comment:
I think that we also need `to_mapping`.

Reply:
I'd prefer not to add it unless we absolutely need it. Nothing should manipulate this data as dictionary (except maybe tests).